### PR TITLE
[Windows] Mono build project restructure dropping .def file enabling full static and dynamic mono builds.

### PR DESCRIPTION
**Pull request by lateralusX**
Major Visual Studio mono project standardization and restructuring in order to drop .def file usage and build a fully merged static libmono library in parallel with current dynamic mono library build.

This PR also continues reorganization of source files into separate targets files previously initialized in eglib's project configuration. The structure and target files follow makefile structure as close as possible. Moving source files into different target files that could be included by different Visual Studio projects making it easier to reused source files and minimize maintenance. The separation of source files into target files have been used in order to do separate builds of dynamic and static mono library on Windows. This was needed in order to drop the .def file previously needed to export symbols when building dynamic mono library.

All symbols decorated with MONO_API will now automatically be exported by the dynamic mono build. Due to the behavior of __declspec (dllexport), we need to build one static version of mono sources as well as a dynamic with different build options (like exporting symbols). These libraries can however build in parallel.

As part of the project restructuring several of the build properties of different vcxproj's have been moved back to project defaults, making it easier to centralize control of build properties in the future.

### CODE REVIEW

**Comment by lateralusX** [comment]
This PR solves https://github.com/mono/mono/pull/4435.

**Comment by lateralusX** [comment]
When moving over to exporting using MONO_API we now export a couple of new methods, previously not exported by the .def files:

mini_get_debug_options
mini_parse_debug_option
mono_build_date
mono_internal_current_level
mono_internal_current_mask
mono_jit_aot_compiling
mono_method_get_reflection_name
mono_opcodes
mono_profiler_enable_coverage
mono_profiler_set_cleanup_callback
mono_profiler_set_method_tail_call_callback
mono_reraise_exception
mono_use_interpreter
mono_use_llvm

They are all marked MONO_API in the code. The following two seems a little suspect due to the internal naming:

mono_internal_current_level
mono_internal_current_mask

Are we OK to export these additional methods?

**Comment by vargaz** [comment]
Some of these are used by xamarin products, hard to say which one. I think its ok to export them on windows if they are exported on unix.

**Comment by alexrp** [comment]
`mono_internal_current_level` and `mono_internal_current_mask` are exported because some of the logging macros exposed in `utils/mono-logger-internals.h` requires them, and these macros are used in the profiler.

File: msvc/Makefile.am [hunk 1]
```diff
@@ -4,14 +4,8 @@ VCXPROJS = $(wildcard *.vcxproj)
 EXTRA_DIST = \
 	     $(VCXPROJS) \
 	     README \
-	     create-windef.pl \
-	     mono.def \
 	     mono.props \
 	     mono.sln \
 	     monoposixhelper.def \
```
Comment (akoeplinger) [inline msvc/Makefile.am]: What about this .def file?

File: msvc/libgc.targets [hunk 2]
```diff
@@ -0,0 +1,181 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ItemGroup Label="libmonogc_la_sources">
+    <ClCompile Include="$(MonoSourceLocation)\libgc\allchblk.c">
+		  <ExcludedFromBuild Condition="'$(MONO_TARGET_GC)'!='boehm'">true</ExcludedFromBuild>
+	  </ClCompile>
```
Comment (akoeplinger) [inline msvc/libgc.targets]: there's some weird mix of tabs/spaces here and below.

File: msvc/libgcmonosgen.targets [hunk 3]
```diff
@@ -0,0 +1,161 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ItemGroup Label="monosgen_sources">
+    <ClInclude Include="$(MonoSourceLocation)\mono\sgen\gc-internal-agnostic.h">
+      <ExcludedFromBuild Condition="'$(MONO_TARGET_GC)'!='sgen'">true</ExcludedFromBuild>
+    </ClInclude>
+    <ClCompile Include="$(MonoSourceLocation)\mono\sgen\sgen-alloc.c">
+		  <ExcludedFromBuild Condition="'$(MONO_TARGET_GC)'!='sgen'">true</ExcludedFromBuild>
+	  </ClCompile>
```
Comment (akoeplinger) [inline msvc/libgcmonosgen.targets]: tabs/spaces

File: msvc/libgc.targets [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    <ClCompile Include="$(MonoSourceLocation)\libgc\alloc.c">
+		  <ExcludedFromBuild Condition="'$(MONO_TARGET_GC)'!='boehm'">true</ExcludedFromBuild>
```
Comment (akoeplinger) [inline msvc/libgc.targets]: I think we could make this nicer to read if we had a property `TargetsSGenGC` or something and then just use:

```
<ExcludedFromBuild>$(TargetsSGenGC)</ExcludedFromBuild>
```

or can we exclude the whole ItemGroup?

**Comment by vargaz** [comment]
Do we still support boehm on windows ? At least we shouldn't support it on win64.

File: mono/mini/mini.h [hunk 5]
```diff
@@ -53,9 +53,11 @@
  * The mini code should not have any compile time dependencies on the GC being used, so the same object file from mini/
  * can be linked into both mono and mono-sgen.
  */
+#if !defined(MONO_DLL_EXPORT) || !defined(_MSC_VER)
```
Comment (vargaz) [inline mono/mini/mini.h]: Why is this needed ?

File: msvc/Makefile.am (hunk 1 above)
Comment (lateralusX) [reply msvc/Makefile.am]: This PR intentionally left that behind since the source was not prepared with __declspec (dllexport) and some of the exported functions was also defined in external headers. This file is not as much updated as the def file for libmono and could be dropped later, if we feel it is worth the effort to fix the source files to get there.

File: msvc/libgc.targets (hunk 2 above)
Comment (lateralusX) [reply msvc/libgc.targets]: I will make a pass over the files and remove some of the editor artifacts.

File: msvc/libgcmonosgen.targets (hunk 3 above)
Comment (lateralusX) [reply msvc/libgcmonosgen.targets]: I will make a pass over the files and remove some of the editor artifacts.

File: msvc/libgc.targets (hunk 4 above)
Comment (lateralusX) [reply msvc/libgc.targets]: Great idea! I will apply that pattern, could probably be used in some other exclude from build cases as well.

**Comment by lateralusX** [comment]
@vargaz Boehm is still supported for 32-bit Windows. The current build configurations has not explicitly limited the use on 64-bit, but it is not supported, maybe we should even disable the possibility to build on 64-bit Windows, excluding all Boehm sources for x64 configuration, then it won't build at all.

File: mono/mini/mini.h (hunk 5 above)
Comment (lateralusX) [reply mono/mini/mini.h]: sgen have some exports in the source files that needs to be exported from dynamic mono library. Since we build all needed sources when building the libmono-dynamic version, we need to set sgen defines for the libmono-dynamic build as well. I didn't want to add defines/file since that will cause maintenance problems so the sgen defines are set for the complete build of libmono-dynamic. The above check then kicks in but is incorrect for this scenario since the build will only use the produced object files when linking libmono-dynamic using specified GC and the object files won't be shared outside the build.

The root problem is the exports from sgen sources, so if we could get away from exporting them from sgen sources and do something similar as already done with other GC exports, we could reuse the objects build in the static version of sgen GC (as we do in the Boehm use case), those objects won't have any __declspec(dllexport) in them since they are build for the static use case.

**Comment by lateralusX** [comment]
Resolved merge conflict due to https://github.com/mono/mono/pull/5517 using rebase.
